pkg/proxy/conntrack/cleanup.go

```diff
@@ -20,95 +20,109 @@ limitations under the License.
+		klog.ErrorS(err, "Failed to list conntrack entries")
+		return
 	}
 
-	if n, err := ct.ClearEntries(ipFamilyMap[ipFamily], filters...); err != nil {
-		klog.ErrorS(err, "Failed to delete stale service connections")
-	} else {
-		klog.V(4).InfoS("Deleted conntrack stale entries for services", "count", n)
+	// The indices "ipIndex" and "portIndex" represent allowed DNAT flows and are
+	// structured as maps for efficient lookups during cleanup. The key of the maps
+	// is origin (original destination [--orig-dst] for ipIndex and
+	// original port [--org-port-dst] for portIndex) and the value is set of allowed
+	// DNATed destinations (--reply-src).
+
+	// ipIndex maps service IPs (ClusterIP, LoadBalancerIPs and ExternalIPs) with the set of endpointIPs.
+	ipIndex := make(map[string]sets.Set[string])
+	// portIndex maps service NodePort with the set of endpointIPs.
+	portIndex := make(map[int]sets.Set[string])
+
+	for svcName, svc := range svcPortMap {
+		// we are only interested in UDP services
+		if svc.Protocol() != v1.ProtocolUDP {
+			continue
+		}
+
+		endpointIPs := sets.New[string]()
+		for _, endpoint := range endpointsMap[svcName] {
+			// Serving endpoints will never black-hole traffic, we consider all
+			// the serving endpoints as allowed DNATed destinations of service.
+			// Changes to the service should not affect existing flows, so we do
+			// not take traffic policies or topology of the service into account.
+			// This ensures that the behavior of UDP services remains consistent
+			// with TCP services.
+			if endpoint.IsServing() {
+				endpointIPs.Insert(endpoint.IP())
+			}
+		}
+
+		// populate the indices
+		ipIndex[svc.ClusterIP().String()] = endpointIPs
+		for _, loadBalancerIP := range svc.LoadBalancerVIPs() {
+			ipIndex[loadBalancerIP.String()] = endpointIPs
+		}
+		for _, externalIP := range svc.ExternalIPs() {
+			ipIndex[externalIP.String()] = endpointIPs
+		}
+		if svc.NodePort() != 0 {
+			portIndex[svc.NodePort()] = endpointIPs
+		}
 	}
-}
 
-// deleteStaleEndpointConntrackEntries takes care of flushing stale conntrack entries related
-// to UDP endpoints. After a UDP endpoint is removed we must flush any conntrack entries
-// for it so that if the same client keeps sending, the packets will get routed to a new endpoint.
-func deleteStaleEndpointConntrackEntries(ct Interface, ipFamily v1.IPFamily, svcPortMap proxy.ServicePortMap, endpointsUpdateResult proxy.UpdateEndpointsMapResult) {
 	var filters []netlink.CustomConntrackFilter
-	for _, epSvcPair := range endpointsUpdateResult.DeletedUDPEndpoints {
-		if svcInfo, ok := svcPortMap[epSvcPair.ServicePortName]; ok {
-			endpointIP := proxyutil.IPPart(epSvcPair.Endpoint)
-			nodePort := svcInfo.NodePort()
-			if nodePort != 0 {
-				filters = append(filters, filterForPortNAT(endpointIP, nodePort, v1.ProtocolUDP))
+	for _, entry := range entries {
+		// we only deal with UDP protocol entries
+		if entry.Forward.Protocol != protocolMap[v1.ProtocolUDP] {
```

Comment:
Is there any way we could have requested just the UDP entries from the kernel?

(Also, you don't need to use `protocolMap` to look up a constant; you can just say `unix.IPPROTO_UDP`.)

Reply:
Filters are not pushed down, atleast in vishwanda/netlink.
https://github.com/vishvananda/netlink/blob/a01829657b53687bcde90c030e2013e482783847/conntrack_linux.go#L199

Even the delete call in `vishwanda/netlink` first does a table dump, matches the flow based on filters and then executes the delete call with raw bytes of the flow retrieved in table dump call.
https://github.com/vishvananda/netlink/blob/main/conntrack_linux.go#L155